---
title: "R Notebook"
output: html_notebook
---

Load packages required for analysis
```{r loading packages}
library(data.table)
library(limma)
library(minfi)
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
library(IlluminaHumanMethylation450kmanifest)
library(RColorBrewer)
library(missMethyl)
library(matrixStats)
library(minfiData)
library(Gviz)
library(DMRcate)
library(stringr)
```


Download the example data
```{r download the example data}
# # the URL for the data download
# url <- "https://ndownloader.figshare.com/files/7896205"
# # download the data
# if(!file.exists("methylAnalysisDataV3.tar.gz")){
# download.file(url, destfile="methylAnalysisDataV3.tar.gz", method="auto")
# }
# # extract the data
# if(!file.exists("./data")){
# untar("methylAnalysisDataV3.tar.gz", exdir=".", compressed="gzip")
# }
```

Get the 450k annotation data
```{r get the 450k annotation data}
ann450k <- getAnnotation(IlluminaHumanMethylation450kanno.ilmn12.hg19)
head(ann450k)
```

Set up data directory
```{r a glance at raw data}
# set up a path to the data directory
dataDirectory <- "/sc/orga/projects/zhangb03a/lei_guo/BM36_methylation_analysis/my_data/methylation_data_processing_example/data"

# list the files
list.files(dataDirectory, recursive=TRUE)
```

Read in the sample sheet for the experiment
```{r read in the sample sheet for the experiment}
# read in the sample sheet for the experiment
targets <- read.metharray.sheet(dataDirectory, pattern="SampleSheet.csv")

targets
```

Read in the raw data from the IDAT files
```{r read in the raw data from the IDAT files}
# read in the raw data from the IDAT files
rgSet <- read.metharray.exp(targets=targets)

rgSet
```

Give the samples descriptive names
```{r give the samples descriptive names}
# give the samples descriptive names
targets$ID <- paste(targets$Sample_Group,targets$Sample_Name,sep=".")

sampleNames(rgSet) <- targets$ID

rgSet
```

QC
```{r QC: calculate the detection p-values}
# calculate the detection p-values
detP <- detectionP(rgSet)

head(detP)
```

```{r QC: remove poor quality samples}
# remove poor quality samples
keep <- colMeans(detP) < 0.05
rgSet <- rgSet[,keep]

# remove poor quality samples from targets data
targets <- targets[keep,]
targets[,1:5]

# remove poor quality samples from detection p-value table
detP <- detP[,keep]
dim(detP)
```

```{r QC: figures and reports}
# examine mean detection p-values across all samples to identify any failed samples
pal <- brewer.pal(8,"Dark2")
par(mfrow=c(1,2))
barplot(colMeans(detP), col=pal[factor(targets$Sample_Group)], las=2,
cex.names=0.8,ylab="Mean detection p-values")
abline(h=0.01,col="red")
legend("topleft", legend=levels(factor(targets$Sample_Group)), fill=pal,
bg="white")
barplot(colMeans(detP), col=pal[factor(targets$Sample_Group)], las=2,
cex.names=0.8, ylim = c(0,0.002), ylab="Mean detection p-values")
legend("topleft", legend=levels(factor(targets$Sample_Group)), fill=pal,
bg="white")

# qc report
qcReport(rgSet, sampNames=targets$ID, sampGroups=targets$Sample_Group, pdf="qcReport.pdf")
```

Normalization
```{r normalization}
# normalize the data; this results in a GenomicRatioSet object
mSetSq <- preprocessQuantile(rgSet)
```

```{r normalization: figures}
# create a MethylSet object from the raw data for plotting
mSetRaw <- preprocessRaw(rgSet)
# visualise what the data looks like before and after normalisation
par(mfrow=c(1,2))
densityPlot(rgSet, sampGroups=targets$Sample_Group,main="Raw", legend=FALSE)
legend("top", legend = levels(factor(targets$Sample_Group)),
text.col=brewer.pal(8,"Dark2"))
densityPlot(getBeta(mSetSq), sampGroups=targets$Sample_Group,
main="Normalized", legend=FALSE)
legend("top", legend = levels(factor(targets$Sample_Group)),
text.col=brewer.pal(8,"Dark2"))
```

Data exploration
```{r data exploration: figure 1}
# MDS plots to look at largest sources of variation
par(mfrow=c(1,2))

plotMDS(getM(mSetSq), top=1000, gene.selection="common", col=pal[factor(targets$Sample_Group)])
legend("top", legend=levels(factor(targets$Sample_Group)), text.col=pal, bg="white", cex=0.7)

plotMDS(getM(mSetSq), top=1000, gene.selection="common", col=pal[factor(targets$Sample_Source)])
legend("top", legend=levels(factor(targets$Sample_Source)), text.col=pal, bg="white", cex=0.7)
```

```{r data exploration: figure 2}
# Examine higher dimensions to look at other sources of variation
par(mfrow=c(1,3))

plotMDS(getM(mSetSq), top=1000, gene.selection="common", col=pal[factor(targets$Sample_Group)], dim=c(1,3))
legend("top", legend=levels(factor(targets$Sample_Group)), text.col=pal, cex=0.7, bg="white")

plotMDS(getM(mSetSq), top=1000, gene.selection="common", col=pal[factor(targets$Sample_Group)], dim=c(2,3))
legend("topleft", legend=levels(factor(targets$Sample_Group)), text.col=pal, cex=0.7, bg="white")

plotMDS(getM(mSetSq), top=1000, gene.selection="common", col=pal[factor(targets$Sample_Group)], dim=c(3,4))
legend("topright", legend=levels(factor(targets$Sample_Group)), text.col=pal, cex=0.7, bg="white")
```

Filtering
```{r filtering poor probes}
# ensure probes are in the same order in the mSetSq and detP objects
detP <- detP[match(featureNames(mSetSq),rownames(detP)),]

# remove any probes that have failed in one or more samples
keep <- rowSums(detP < 0.01) == ncol(mSetSq)
table(keep)

mSetSqFlt <- mSetSq[keep,]
mSetSqFlt
```

```{r filtering probes on the sex chromosomes}
# if your data includes males and females, remove probes on the sex chromosomes
keep <- !(featureNames(mSetSqFlt) %in% ann450k$Name[ann450k$chr %in% c("chrX","chrY")])
table(keep)

mSetSqFlt <- mSetSqFlt[keep,]
```

```{r remove probes with SNPs at CpG site}
# remove probes with SNPs at CpG site
mSetSqFlt <- dropLociWithSnps(mSetSqFlt)

mSetSqFlt
```

```{r exclude cross reactive probes}
# exclude cross reactive probes
xReactiveProbes <- read.csv(file=paste(dataDirectory, "48639-non-specific-probes-Illumina450k.csv", sep="/"), stringsAsFactors=FALSE)
keep <- !(featureNames(mSetSqFlt) %in% xReactiveProbes$TargetID)
table(keep)

mSetSqFlt <- mSetSqFlt[keep,]
mSetSqFlt
```

Re-examine the MDS plots to see if the relationship between the samples has changed after filtering
```{r re-examine the MDS plots}
par(mfrow=c(1,2))

plotMDS(getM(mSetSqFlt), top=1000, gene.selection="common", col=pal[factor(targets$Sample_Group)], cex=0.8)
legend("right", legend=levels(factor(targets$Sample_Group)), text.col=pal, cex=0.65, bg="white")

plotMDS(getM(mSetSqFlt), top=1000, gene.selection="common", col=pal[factor(targets$Sample_Source)])
legend("right", legend=levels(factor(targets$Sample_Source)), text.col=pal, cex=0.7, bg="white")
```

```{r re-examine higher dimensions to look at other sources of variation}
# Examine higher dimensions to look at other sources of variation
par(mfrow=c(1,3))

plotMDS(getM(mSetSqFlt), top=1000, gene.selection="common", col=pal[factor(targets$Sample_Source)], dim=c(1,3))
legend("right", legend=levels(factor(targets$Sample_Source)), text.col=pal, cex=0.7, bg="white")

plotMDS(getM(mSetSqFlt), top=1000, gene.selection="common", col=pal[factor(targets$Sample_Source)], dim=c(2,3))
legend("topright", legend=levels(factor(targets$Sample_Source)), text.col=pal, cex=0.7, bg="white")

plotMDS(getM(mSetSqFlt), top=1000, gene.selection="common", col=pal[factor(targets$Sample_Source)], dim=c(3,4))
legend("right", legend=levels(factor(targets$Sample_Source)), text.col=pal, cex=0.7, bg="white")
```


Calculate M-values for statistical analysis
```{r calculate M and Beta values}
# calculate M-values for statistical analysis
mVals <- getM(mSetSqFlt)
head(mVals[,1:5])

# calculate Beta-values for statistical analysis
bVals <- getBeta(mSetSqFlt)
head(bVals[,1:5])
```

```{r plot M and Beta values}
par(mfrow=c(1,2))

densityPlot(bVals, sampGroups=targets$Sample_Group, main="Beta values", legend=FALSE, xlab="Beta values")
legend("top", legend = levels(factor(targets$Sample_Group)), text.col=brewer.pal(8,"Dark2"))

densityPlot(mVals, sampGroups=targets$Sample_Group, main="M-values", legend=FALSE, xlab="M values")
legend("topleft", legend = levels(factor(targets$Sample_Group)), text.col=brewer.pal(8,"Dark2"))
```


Write M and Beta values
```{r write M and Beta values}
fwrite(as.data.frame(mVals), "M_Values_qced_normlized.tsv", sep = "\t", row.names = T)

fwrite(as.data.frame(bVals), "Beta_Values_qced_normlized.tsv", sep = "\t", row.names = T)
```

